Fixes #2787 by checking allowed cron users

pgorod opened the pull request:
This PR (finally) fixes #2787.


## Motivation and Context

### Why we should care
I see this in the forums almost every week. When releases come out and people start upgrading, this problem shows several times a week. I would say all the years we've been suffering from this, and the hours of work consumed by admins and support, all add up to a huge and senseless pain.

### Aren't educational efforts enough?
Sure, people should know how to configure things correctly. However, we're not providing instructions for this in the wiki, there are tons of wrong instructions out there on the Internet, and our installer misleads people.
No, educational efforts are not enough - they are _what we have been failing at_ for all these years. No matter how often I explain all this in the forums, it comes up again and again.

### Why it's not that simple
Because configurations can be very different from system to system, and we need to accomodate them all. Because it's hard to know from within cron, which user your other scripts (the web app) run under. So this information needs to be recorded in one place, then read in another.

## Description

### What it does, and how it impacts people
- It does not affect Windows at all. This is a problem of permissions degradation caused by cron misconfiguration, it was diagnosed in Linux, and the fix affects Linux only (in case you're wondering, `exec('whoami');` works on Windows)
- new installations on Linux will get proper advice as to how to setup cron correctly for their specific case.
- On upgrades, it does not affect people with standardly configured Linux systems (web server user = cron user). These should just go on working.
- On upgrades, people with non-standard Linux configurations will have to review their config.php to keep cron working. They will get explanative log entries and receive instructions about this in the Release notes.

### Technical details
- **on install time**, add an entry to an array in `config.php` called `allowed_cron_users` containing the current web server user. 
- **on install time**, provide correct instructions for cron configuration with that specific user's crontab.
- **on upgrade time**, add the same entry to the same array, if it's not there yet (for people coming from versions of SuiteCRM before this mechanism was introduced)
- for these automatic additions of users to the array, **`root` is treated as a special case**, to discourage its use and to force manual intervention from an admin
- **on Schedulers screen**, show specific correct configuration instructions also
- **when `cron.php` runs**, check if the current user is in the array, and exit if not, leaving appropriate log message 

### Security
Any attack that compromises `cron.php` _or any of the Scheduler scripts that it runs_ will be much more severe if cron is running as `root`, something that happens in a lot of installations out there. We're doing people a favor by nudging them towards a safer configuration.

### Educational efforts

As I said, these are not enough, but they _are_ needed.

#### Suggested text snippet to include in Release Notes:

> New mechanism to check which user cron runs under, to avoid permissions degradation issues. IMPORTANT: after upgrade, check your logs and confirm that your Schedulers are running.


#### Suggested addition to the [wiki installation instructions](https://suitecrm.com/wiki/index.php/Installation):

Insert the following section just before the Title "Upgrading SuiteCRM". This will make it the final section of the previous title ("Installing SuiteCRM").

> "Post-Installation steps"
> 
> To complete installation, the Scheduler jobs need to be enabled. This is done by asking your operating system to run a script called cron.php every minute. This, in turn, manages all SuiteCRM jobs that are supposed to run at scheduled times, supporting functionalities like search indexing, workflows, email notifications, database maintenance, etc.
> 
> In Windows this is setup on the system's "Task Scheduler"; in Linux and iOS, it is setup in crontab. During the installation you will receive detailed instructions with the specific commands you need to run to setup these jobs. After installation, you can still access these instructions in Admin area, under option "Schedulers", at the bottom of the screen.
> 
> To make sure your Scheduler jobs don't create permissions problems, the standard practice is to run them under the same user that your Web server runs under. For different configurations, the allowed_cron_users array in config.php must be adjusted.
> 

#### In the forums

Of course some people will get into trouble with this when upgrading and show up in the forums saying "my cron jobs aren't working", and we'll simply answer: complete your upgrade by following instructions, here is the link to configure cron correctly. We have an admin's attention, he scheduled downtime for this, he's probably even testing on a clone of his production system.

Compare this with what we have now: people come up weeks or months after creating the problem, with complains from end-users and managers, with random symptoms and sometimes varied and extensive system rot in production. After several diagnosis attempts, they are told to chmod their permissions, and the cycle of permissions' degradation starts over, to need fixing again later on...

## How To Test This
I tested these situations:
1. Installation (check `config.php` gets a correct `cron_allowed_users` entry).
2. Upgrade (to check the same as above. I had to create a tweaked upgrade package with the new files).
3. Running cron, both with an allowed user and a non-allowed user.
4. You can also try some variations on the user (root or non-root), etc

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ X ] Bug fix (non-breaking change which fixes an issue)
- [ X ] New feature (non-breaking change which adds functionality)
- [ X ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
<!--- Go over all the following points and check all the boxes that apply. --->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! --->
- [ X ] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ X ] **My change requires a change to the documentation (text suggestions provided).**
- [ X ] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php
/Contributing_to_SuiteCRM) guidelines.

horus68 commented:
Language strings added to translations packs for versions 7.8.2.3
Just ping me if there are any language changes in this PR

pgorod commented:
@horus68 you add the strings _before_ the PR's get merged? Doesn't that make things complicated for you when the merges are delayed for a long time?

horus68 commented:
No... if I just include the new strings and let the old ones stay around.
This alow to test the PR with translation but to work on previous versions.
This also gives translators the time to translate. Then when a PR is merged its ready to be used by all!

chris001 commented:
I have a similar fix for this issue, testing it now.

anothermouse commented:
Just looking at some of the shared hosting environments that don't directly allow shell commands, or run in a chrooted environment. Perhaps:
     posix_getpwuid(posix_geteuid())['name'];
is a better alternative to exec('whoami'); ?

pgorod commented:
People on those systems could still configure the user manually in `config.php`, but I agree it would be better if we could handle things automatically everywhere. I'd like to avoid diverging too much from Windows, though. Anything in the code related to permissions already diverges _a lot_. We're basically always assuming Windows is entirely permissive which, of course, is not true everywhere, and I believe will tend to be less so in the future.

Anyway, thanks @anothermouse because your comment is obviously something I need to look at. Do you know of any place where I can test those cases of shared hostings with those limitations? I'd like to see exactly how `exec` fails and how I could test for that.

pgorod commented:
Ok, I got it - if I add `exec` to `disabled_functions` in my `php.ini` I can emulate does scenarios. Testing now.

pgorod commented:
I added some more logic, it's probably overkill, but it might come in useful in some weird hosting out there. Anyway it now works on systems with `exec` disabled.

samus-aran commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Robustness Big Issue 1: cron shouldn't be allowed to run as root](https://trello.com/c/wTto2LbF/48-robustness-big-issue-1-cron-shouldn-t-be-allowed-to-run-as-root)

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+/**
+ * Gets the username of the user under which the PHP script is currently running
+ * Notes:
+ * - works on Windows and Linux, tries a variety of methods to accomodate different systems and hosting restrictions
```
typo accommodate

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+
+function getRunningUser()
+{
+    // works on Windows and Linux, but might return null on systems that include exec in disabled_functions in php.ini (typical in shared hosting)
```
exceeds 120 characters, break into 2 lines

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+ * Notes:
+ * - this is Linux only, does nothing on Windows
+ * - does not repeat the user if he is already there
+ * - creates the sub-array if previoulsy unexisting
```
typo previously

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+            $addUser = 'root_REMOVE_THIS_NOTICE_IF_YOU_REALLY_WANT_TO_ALLOW_ROOT';
+            if (!in_array($addUser, $sugar_config['cron']['allowed_cron_users'])) {
+                $sugar_config['cron']['allowed_cron_users'][] = $addUser;
+                $GLOBALS['log']->error("You're using 'root' as the web-server user. This should be avoided for security reasons. Review allowed_cron_users configuration in config.php.");
```
exceeds 120 characters, break into 2 lines

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+        else
+            $runningUser = posix_getpwuid(posix_geteuid())['name'];
+    }
+    return ($runningUser == null) ? "" : $runningUser;
```
use single quotes

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+{
+    global $sugar_config;
+
+    if (is_windows() || !isset($sugar_config)|| !isset($addUser) || ($addUser == "")) {
```
use single quotes

mattlorimer commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+ * - special treatment for user 'root' to require manual intervention from an admin to allow
+ */
+
+function addCronAllowedUser(string $addUser)
```
scalar type hints e.g string, are not supported in php < 7

anothermouse commented on install/install_utils.php:
```diff
@@ -2169,3 +2169,66 @@ function enableInsideViewConnector()
+    $runningUser = exec('whoami');
+
+    if ($runningUser == null) {  // matches null, false and ""
+        if (is_windows())
```
Can we change line 2184 to use the getRunningUser() function for consistency?

pgorod replied:
@anothermouse this is _inside_ getRunningUser, so that would cause an endless recursion. This function uses both methods, whoami and the one you suggested, it's probably overkill but my idea was to cover as many different scenarios as possible (where one function works, but not the other, and vice-versa).

anothermouse replied:
Sorry - my mistake....I didn't examine it closely enough as I was only looking at the diffs. I was thinking about the uniformity of using:
posix_getpwuid(posix_geteuid())['name'];
and avoiding the use of exec('whoami')

pgorod commented:
@mattlorimer please confirm if I understood you correctly: I moved both functions, `getRunningUser` and `addCronAllowedUser` to `include/utils.php`.

samus-aran commented:
@pgorod - Yes that is what he meant so we can use it else where when required.

Couple of things on my side and its regarding testing which I have only really got round to doing. Anyway - when installing.  The amendments to actually add the allowed_cron_users to the config does not execute because complete_install.php never gets called.  The last step within the SuiteCRM install process is 'performSetup.php' so I would suggest putting your functionality in there.  We have two places but I recommend the latter - I'm trying to find the best place myself.

Now I was going to look at this today and push up to your branch but I wanted to make sure I wasn't missing something extremely obvious as to why my allowed_cron_users were not appearing in my config.php

Reason why I noticed it was when I was running the cron, I expected an error as there was not anything in config.php but it didn't as the check expects it to be there... in which case we should put an error log to notify that there is not allowed_cron_users set up. https://github.com/Accendere/SuiteCRM/blob/manage-web-server-users/cron.php#L56

Another minor thing is to update to where the file to be pointing. https://github.com/Accendere/SuiteCRM/blob/manage-web-server-users/cron.php#L53

pgorod commented:
Hi @samus-aran thanks for the tips.

I fixed the `require` in `cron.php`.

I added a warning in `cron.php` when no `allowed_cron_users` entry is found in `config.php`.

Regarding the code in `install.php`, I did test that (quite a while ago - maybe I need to try it again) and I believe the code ran. Maybe `complete_setup.php` is not getting required as a file, but it seems the ` case 'complete_install.php':` section in `install.php` is running. 

It just occurred to me - you don't happen to be using a silent install method? I didn't test that (and probably should). Anyway the place we choose to put that code in, should work for both kinds of installs...

samus-aran commented:
Thanks @pgorod - I would recommend again to test it as I've just cloned your repo, checkout manage-web-server-users and ran install.php (via the browser, not silent install) and allowed_cron_users are not being inserted into config.php.  My only thinking is that this has changed since you tested (due to our delay on picking this up) but I think putting it in performSetup would actually be more beneficial as complete_install is pretty redundant (it only cleared the session and redirected you to login page which performSetup does).

My steps were:
- clone https://github.com/Accendere/SuiteCRM
- checkout manage-web-server-users
- chown www-data:www-data -R .
- chmod 775 -R . (yeah abit lazy, but for quickness)
- navigated to url/install.php and ran through installation wizard, completed and directed to login page.
- open newly created config.php - search for allowed_cron_users (not found).

pgorod commented:
Ok I did the changes, I agree it looks better this way, and probably gets this working also for silent installs.

I can't test right now (no access to my installation tests VM on a Sunday) but I hope to do so soon. Anyway I pushed the changes so that if you get a chance, you can test it. Thanks.

samus-aran commented:
Thanks @pgorod - Tested that and it works - Superduper.  I'll do a couple of other tests but overall very happy with that.

pgorod commented:
Great!  :-)

The test that gave me more work was the upgrade. I had to install an earlier version (7.8.1), then get an upgrade package for 7.8.2 and tweak it manually, inserting my files into the zip.

I was afraid that my new file `end.php` would not be in place at the end _of that same upgrade_, which was required. But thankfully it was copied somewhere in the middle of the upgrade, and was in place when the `end.php` was called.

pgorod commented:
Thanks @samus-aran!

Don't forget to copy-paste the texts above for the Release Notes and the Wiki (changing whatever you feel needs changing).
